### DISCUSSION TITLE
Feature Idea: Java AppCDS

### DISCUSSION THREAD

**Discussion by ennerf** [hydraulic-software/conveyor#95] [Ideas]
One of the reasons for Java's notoriously slow startup is that each JVM instance loads and compiles classes separately every time. To mitigate some of that, [JEP 310](https://openjdk.org/jeps/310) (JDK10) introduced Application Class-Data Sharing (AppCDS), which allows instances to share pre-loaded class data. In some initial tests I saw about 20-40% lower startup time for different JavaFX applications.

I'm thinking about ways to enable it for users, but I think the best place for it would be at the Conveyor level. The shared class data is specific to the user machine, so it'd have to be generated locally into a persistent directory and get deleted after an update.

The steps needed to do this on jdk21 are

1. create a static list of classes that are used in the application. This can be done locally before release and might require some dev interaction (e.g. clicking different dialogs so all classes get listed). `-XX:DumpLoadedClassList=app-class-list.txt` (and `-Xlog:cds` for logging)
2. compile the class list locally. This would need to be done after installation or before the first run. It's unfortunately a fairly lengthy step and can take ~20s on my machine. `-Xshare:dump -XX:SharedClassListFile=app-class-list.txt -XX:SharedArchiveFile=app-cds.jsa`. 
3. start the application with the shared archive file `-Xshare:on -XX:SharedArchiveFile=app-cds.jsa`

**Comment by mikehearn** [comment]
We've looked at this. You should be able to ship an archive as a data file with the app, at least in theory.

There are some issues we hit whilst experimenting with something better:

1. When run in dynamic dump mode the JVM gets super slow. This would be the first time the user runs the app or after any upgrade, so it would be a bad first impression. If the user runs a long session after an upgrade, performance would be poor.
2. You need some extra code on the C++ side to do things like verify the user has enough disk space, delete old archives created by old versions of the app, trigger re-creation on upgrade/downgrade and so on.
3. We have had to disable AppCDS internally sometimes due to experiencing weird bugs like class verification errors at startup, etc.
4. On macOS there's no way to clean up old caches when the user uninstalls an app and they can get quite large. That's more of a macOS issue and it goes away if you opt into sandboxing (weirdly enough), but it's one more thing to address.

Given the odd flakiness I'm not sure anyone is using AppCDS for real, so it'd be a bit of a risk to enable it.

Overall it's normally a ~30% startup time win, so not nothing, but also kind of on the borderline for what's worth it.

Still, it has the advantage of working with everything out of the box, and they're improving it with Leyden. What I'm hoping is that they fix the speed of dynamic dumping and solve the workflow issues with deleting/recreating the archives so there's no need for Conveyor to do that.

**Reply by ennerf** [reply]
I had issues with dynamic mode as well, but I'm thinking the static way might be easier. The flakiness and weird bugs are probably a showstopper for now though.

**Comment by rcoleb** [comment]
I know there are issues beyond what I've encountered (I haven't experienced the class verification startup errors, for example).  
No idea if any of this is new info, but hoping this might be helpful, I have a couple of notes / thoughts on AppCDS usage:

 * You can specify the location of both the static and dynamic CDS files - if these locations remain the same across updates, archives will be recreated / replaced, rather than accruing per-version archives.  
 
 * I've had success with managing user reception by incorporating CDS creation as part of the install / update process; this pushes the bulk of the slowdown to a process that users (in my experience) are often more accustomed to taking longer.
 
   * This is more easily achievable for any Java application when creating the static CDS archive, as Java automatically exits after creating the archive. Prior to using Conveyor, I found it fairly straightforward to create a second launcher with JPackage with the `-Xshare:dump` flag specified, and then run that launcher as a post-install step (using an InnoSetup-created installer).  I recognize that the Conveyor documentation, in the Launch4J comparison section, says, "Conveyor builds packages rather than installers" - I don't know if this is still considered true given that Conveyor does indeed create installers?
   
   * The application I am working on uses Spring Boot, and I believe Spring is part of why I've seen startup time improvements >50% - I suspect a good part of this is due to JPA and AOP proxies.
   
     * To create the Dynamic CDS, I have been using `-Dspring.context.exit=onRefresh` to load and instantiate all Spring-required classes and then exit automatically. This is packaged in another JPackage-created launcher and run as a post-install step after the static CDS creation completes. 

 * I have found that the static list of classes is not necessary. The options / steps I've implemented are:
   _Static CDS:_  
  `-Xshare:dump`  
  _Dynamic CDS:_
  `-XX:ArchiveClassesAtExit=<path\to\>custom-cds.jsa -Dspring.context.exit=onRefresh`
  _Application:_
  `-XX:+AutoCreateSharedArchive -XX:SharedArchiveFile=<path\to\>custom-cds.jsa`

  The combination of creating the Dynamic CDS based on Spring's classes, coupled with the `-XX:+AutoCreateSharedArchive` flag, results in both significant initial startup times, as well as minor ongoing speed improvements as users use specific features or components that aren't immediately instantiated by Spring on startup. 

Even if Conveyor had a Spring-specific Dynamic CDS option, I wonder if the prevalence of Spring would make this a worthwhile feature. And certainly all Java apps could use the static CDS.

**Reply by mikehearn** [reply]
Thanks, this is a really good writeup!

Conveyor can run arbitrary code post-install on Windows _sometimes_ (users aren't obliged to use the installer EXE), but AppCDS is an optimization so that's not necessarily an issue. Still, users don't like to wait for an install and we do try to keep it fast. Debian packages can run code. The issue is more macOS where there's genuinely no installer. Better support for Native Image might be the way to go as then the download itself is pre-optimized.
